processor/k8sattributesprocessor/internal/kube/kube.go

```diff
@@ -169,6 +171,56 @@ type FieldExtractionRule struct {
 	From string
 }
 
+func (r *FieldExtractionRule) extractFromPodMetadata(metadata map[string]string, tags map[string]string, formatter string) {
+	// By default if the From field is not set for labels and annotations we want to extract them from pod
+	if r.From == MetadataFromPod || r.From == "" {
+		r.extractFromMetadata(metadata, tags, formatter)
+	}
+}
+
+func (r *FieldExtractionRule) extractFromNamespaceMetadata(metadata map[string]string, tags map[string]string, formatter string) {
+	if r.From == MetadataFromNamespace {
+		r.extractFromMetadata(metadata, tags, formatter)
+	}
+}
+
+func (r *FieldExtractionRule) extractFromMetadata(metadata map[string]string, tags map[string]string, formatter string) {
+	if r.KeyRegex != nil {
+		for k, v := range metadata {
+			var name string
+			if r.HasKeyRegexReference {
```

Comment:
We can drop this condition along with the `else` block, if we have `formatter` as a string with regexp caption e.g. `k8s.namespace.annotations.$0`.

Reply:
It seems to fully leverag `$0` it has certain requirements for the regexp, so far basically the difference is how it behaviors between `*` and `.*`, e.g. https://go.dev/play/p/yUbTV_SKWCj

Reply:
Sorry I don't understand your point. Can you elaborate please

Reply:
$0 doesn't always represent the entire key, i.e. in https://go.dev/play/p/yUbTV_SKWCj, given the key is `annotation1`, name is `annotations.$0`,
* given the key_regex is `an*`, $0 would be `ann` and `a`(both matches the regexp), after 2 iterations(the for loop in the sample code), we got the result `annotations.annannotations.a` (`annotations.ann` + `annotations.a`).
* given the key_regex is `an.*`, $0 would be `annotation1` as the matched string is the entire key, thus just one iteration and we got the expected result `annotations.annotation1`

Reply:
This seems like an incorrect expansion logic. the result should not be composition of several matches.

Reply:
updated this part to perform substitution only once, example https://go.dev/play/p/CtD2RwehIWm

I think it's a different behavior in golang regexp implementation. Submatch 0 is **the match of** the entire expression as per [regexp doc](https://pkg.go.dev/regexp#pkg-overview), but not the entire string that's supplied. 

> If 'Submatch' is present, the return value is a slice identifying the successive submatches of the expression. Submatches are matches of parenthesized subexpressions (also known as capturing groups) within the regular expression, numbered from left to right in order of opening parenthesis. Submatch 0 is the match of the entire expression, submatch 1 is the match of the first parenthesized subexpression, and so on.

And also in Python, group 0 means the same. 

```
$ python3 -c 'import re; print(re.search(r"(an*)","annotation1").expand(r"\g<0>"))'
ann

$ python3 -c 'import re; print(re.search(r"an*","annotation1").expand(r"\g<0>"))'
ann
```

IMO we should keep this behavior consistent with regexp package to not surprise the users but add some notes about this, what do you think?

Reply:
> updated this part to perform substitution only once, example https://go.dev/play/p/CtD2RwehIWm

I am not sure what's exactly the concern. I think (but might be wrong) nobody would write the pattern as `an*`, this means ["zero or more 'n'"](https://pkg.go.dev/regexp/syntax) in this case. For `an.*` (as well as `(an.*)`) all three examples behave the same

Having said that, I think "substitute2" is providing the sort of output I would be anticipating here

Reply:
As far as I can tell, this is the current implementation so I think we are good?

Reply:
I think we should do strict full-length match on the provided attribute value. I don't believe `n.*` shouldn't match against `annotation1`, but `an.*`, `.*` etc should.  Just wrapping with `^` and `$` whatever user supplied should do the trick.

Reply:
I think this thread was raised to see if the if else block can be removed by leveraging `k8s.namespace.annotations.$0`. 

I think we can't because it would be a breaking change, say user just set `key_regex` to `an*`, it would produce `k8s.namespace.annotations.ann` while previously it is `k8s.namespace.annotations.annotation1`.

Reply:
> I think we can't because it would be a breaking change, say user just set `key_regex` to `an*`, it would produce `k8s.namespace.annotations.ann` while previously it is `k8s.namespace.annotations.annotation1`.

Are you sure that currently `an*` matches `annotation1` annotation and results into `k8s.namespace.annotations.annotation1`? If that's the case, we can probably keep the else block. But I still believe that we should do strict match and make `an*` do not match `annotation1`, only `annnnnn`. @pmm-sumo if you agree on this behavior, I believe we can introduce this breaking change as another PR.

Reply:
> Just wrapping a regexp supplied by user with ^ and $should do the trick

would it be tricky to wrap it with ^ and $, given that user may have ^ or $ in their regexp, ^ and $ probably has to be appended during validation phase, where we need to handle cases like
* ^an.*
* ^an.*$
* (^an.*)
* (^an.*$)

Reply:
> would it be tricky to wrap it with ^ and $, given that user may have ^ or $ in their regexp, ^ and $ probably has to be appended during validation phase, where we need to handle cases like

Yes, this would be the right approach. But let's not do it for now

Reply:
@dmitryax yeah the reason I noticed that `an*` is while making this change and noticed failures from current UT, which is expecting `k8s.namespace.annotations.annotation1`. https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/29dd41245f5a2e628d7bdc9ac33f94637645a000/processor/k8sattributesprocessor/internal/kube/client_test.go#L650-L663

Reply:
>  But I still believe that we should do strict match and make an* do not match annotation1, only annnnnn. @pmm-sumo if you agree on this behavior, I believe we can introduce this breaking change as another PR.

Yeah, I have the same sentiment @dmitryax @newly12

Reply:
Created a ticket https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9716. @newly12 feel free to pick it up if you are interested